src/infra/core/src/use_cases/create_dataset_from_snapshot_use_case_impl.rs

```diff
@@ -30,29 +34,69 @@ use crate::DatasetRepositoryWriter;
 #[interface(dyn CreateDatasetFromSnapshotUseCase)]
 pub struct CreateDatasetFromSnapshotUseCaseImpl {
     current_account_subject: Arc<CurrentAccountSubject>,
+    dataset_repo_reader: Arc<dyn DatasetRepository>,
     dataset_repo_writer: Arc<dyn DatasetRepositoryWriter>,
+    rebac_service: Arc<dyn RebacService>,
     outbox: Arc<dyn Outbox>,
 }
 
 impl CreateDatasetFromSnapshotUseCaseImpl {
     pub fn new(
         current_account_subject: Arc<CurrentAccountSubject>,
+        dataset_repo_reader: Arc<dyn DatasetRepository>,
         dataset_repo_writer: Arc<dyn DatasetRepositoryWriter>,
+        rebac_service: Arc<dyn RebacService>,
         outbox: Arc<dyn Outbox>,
     ) -> Self {
         Self {
             current_account_subject,
+            dataset_repo_reader,
             dataset_repo_writer,
+            rebac_service,
             outbox,
         }
     }
+
+    async fn handle_multi_tenant_case(
+        &self,
+        create_dataset_result: &CreateDatasetResult,
+        options: &CreateDatasetFromSnapshotUseCaseOptions,
+    ) -> Result<(), CreateDatasetFromSnapshotError> {
+        // Trying to set the ReBAC property
+        let allows = options.dataset_visibility.is_publicly_available();
+        let (name, value) = DatasetPropertyName::allows_public_read(allows);
+
+        let set_rebac_property_res = self
+            .rebac_service
+            .set_dataset_property(&create_dataset_result.dataset_handle.id, name, &value)
+            .await
+            .map_err(|err| match err {
+                SetEntityPropertyError::Internal(e) => CreateDatasetFromSnapshotError::Internal(e),
+            });
+
+        // If setting fails, try to clean up the created dataset
+        if let Err(set_rebac_property_err) = set_rebac_property_res {
+            self.dataset_repo_writer
+                .delete_dataset(&create_dataset_result.dataset_handle)
+                .await
+                // Return a deletion error if there is one
+                .map_int_err(CreateDatasetFromSnapshotError::Internal)?;
+
+            // On successful deletion, return the property setting error
+            return Err(set_rebac_property_err);
+        }
+
```

Comment:
I think there might be a better approach now using the Outbox.
1) Make uniqueness validation before doing anything. I don't think it's possible to break invairants of ReBac repository like name uniqueiness, if the same check has already passed in the storage repository.

2) Create a dataset in the storage part (disk or S3). If it fails, the entire scenario fails.

3) There already is a message issued into outbox about a created dataset,. It can be extended with extra attributes like publicity, if necessary.

4) Re-Bac part may subscribe to event and create a necessary record. If this process fails from technical issues, like DB connectivity, outbox will ensure retrial until transaction passes.

It would be really beneficial of `core` domain was complete unaware of `rebac` domain.